File: src/main/java/com/zerone/math/Vector2.java
package com.zerone.math;

public class Vector2 {

    public float x;
    public float y;

    public Vector2() {

    }

    public Vector2(float x, float y) {
        this.x = x;
        this.y = y;
    }

    public Vector2(Vector2 other) {
        this.x = other.x;
        this.y = other.y;
    }

    public Vector2 cpy() {
        return new Vector2(x, y);
    }

    public Vector2 set(float x, float y) {
        this.x = x;
        this.y = y;
        return this;
    }

    public Vector2 set(Vector2 other) {
        this.x = other.x;
        this.y = other.y;
        return this;
    }

    public Vector2 add(float x, float y) {
        this.x += x;
        this.y += y;
        return this;
    }

    public Vector2 add(Vector2 other) {
        this.x += other.x;
        this.y += other.y;
        return this;
    }

    public Vector2 sub(float x, float y) {
        this.x -= x;
        this.y -= y;
        return this;
    }

    public Vector2 sub(Vector2 other) {
        this.x -= other.x;
        this.y -= other.y;
        return this;
    }

    public Vector2 mul(float scalar) {
        this.x *= scalar;
        this.y *= scalar;
        return this;
    }

    public float len() {
        return Mathf.sqrt(x * x + y * y);
    }

    public Vector2 nor() {
        float len = len();
        if (len != 0) {
            this.x /= len;
            this.y /= len;
        }
        return this;
    }

    public float angle() {
        float angle = Mathf.toDegrees(Mathf.atan2(y, x));
        if (angle < 0)
            angle += 360;
        return angle;
    }

    public Vector2 rotate(float angle) {
        float rad = Mathf.toRadians(angle);
        float cos = Mathf.cos(rad);
        float sin = Mathf.sin(rad);

        float newX = this.x * cos - this.y * sin;
        float newY = this.x * sin + this.y * cos;

        this.x = newX;
        this.y = newY;

        return this;
    }

    public float dist(Vector2 other) {
        float distX = this.x - other.x;
        float distY = this.y - other.y;
        return Mathf.sqrt(distX * distX + distY * distY);
    }

    public float dist(float x, float y) {
        float distX = this.x - x;
        float distY = this.y - y;
        return Mathf.sqrt(distX * distX + distY * distY);
    }

    public float distSquared(Vector2 other) {
        float distX = this.x - other.x;
        float distY = this.y - other.y;
        return distX * distX + distY * distY;
    }

    public float distSquared(float x, float y) {
        float distX = this.x - x;
        float distY = this.y - y;
        return distX * distX + distY * distY;
    }

}

File: src/main/kotlin/com/zerone/math/Vector2.kt
package com.zerone.math

data class Vector2(var x: Float = 0.toFloat(), var y: Float = 0.toFloat()) {

    constructor(other: Vector2): this(other.x, other.y)

    fun set(other: Vector2): Vector2 {
        this.x = other.x
        this.y = other.y
        return this
    }

    fun set(x: Float, y: Float): Vector2 {
        this.x = x
        this.y = y
        return this
    }

    fun plus(x: Float, y: Float): Vector2 {
        this.x += x
        this.y += y
        return this
    }

    fun plus(other: Vector2): Vector2 {
        this.x += other.x
        this.y += other.y
        return this
    }

    fun sub(x: Float, y: Float): Vector2 {
        this.x -= x
        this.y -= y
        return this
    }

    fun sub(other: Vector2): Vector2 {
        this.x -= other.x
        this.y -= other.y
        return this
    }

    fun times(scalar: Float): Vector2 {
        this.x *= scalar
        this.y *= scalar
        return this
    }

    fun len(): Float {
        return Mathf.sqrt(x * x + y * y)
    }

    fun nor(): Vector2 {
        val len = len()
        if (len != 0f) {
            this.x /= len
            this.y /= len
        }
        return this
    }

    fun angle(): Float {
        var angle = Mathf.toDegrees(Mathf.atan2(y, x))
        if (angle < 0)
            angle += 360f
        return angle
    }

    fun rotate(angle: Float): Vector2 {
        val rad = Mathf.toRadians(angle)
        val cos = Mathf.cos(rad)
        val sin = Mathf.sin(rad)

        val newX = this.x * cos - this.y * sin
        val newY = this.x * sin + this.y * cos

        this.x = newX
        this.y = newY

        return this
    }

    fun dist(other: Vector2): Float {
        val distX = this.x - other.x
        val distY = this.y - other.y
        return Mathf.sqrt(distX * distX + distY * distY)
    }

    fun dist(x: Float, y: Float): Float {
        val distX = this.x - x
        val distY = this.y - y
        return Mathf.sqrt(distX * distX + distY * distY)
    }

    fun distSquared(other: Vector2): Float {
        val distX = this.x - other.x
        val distY = this.y - other.y
        return distX * distX + distY * distY
    }

    fun distSquared(x: Float, y: Float): Float {
        val distX = this.x - x
        val distY = this.y - y
        return distX * distX + distY * distY
    }

    override fun toString(): String {
        return "$x, $y"
    }
}
